File: src/me/tcklpl/naturaldisaster/disasters/Thunderstorm.java
package me.tcklpl.naturaldisaster.disasters;

import me.tcklpl.naturaldisaster.map.DisasterMap;
import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.BlockFace;
import org.bukkit.entity.Creeper;
import org.bukkit.entity.Entity;
import org.bukkit.entity.LightningStrike;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;

public class Thunderstorm extends Disaster {

    public Thunderstorm(DisasterMap map, JavaPlugin main) {
        super(map, main);
        name = "Thunderstorm";
        hint = "Evite locais altos.";
    }

    @Override
    public void startDisaster() {
        super.startDisaster();

        map.makeRain(false);

        Random r = random;

        int strikechance = 10;

        taskId = Bukkit.getScheduler().scheduleSyncRepeatingTask(main, () -> {

            HashMap<Location, Integer> creeperChargedSpawns = new HashMap<>();

            for (String playerName : map.getPlayersInArena()) {

                Player p = Bukkit.getPlayer(playerName);
                assert p != null;
                Location l = p.getLocation();

                if (p.getGameMode() == GameMode.ADVENTURE) {

                    int chanceOfStrike = 0;
                    int playersNearby = 0;

                    // Lightning chance of spawning per factor
                    if (l.getY() >= Math.floorDiv(map.top - map.floor, 2))
                        chanceOfStrike += 30;
                    if (map.getWorld().getHighestBlockYAt(l.getBlockX(), l.getBlockZ()) <= l.getBlockY())
                        chanceOfStrike += 80;
                    for (Entity nearby : p.getNearbyEntities(5, 2, 5)) {
                        if (nearby instanceof Player) {
                            playersNearby += 1;
                            chanceOfStrike += 10;
                        }
                    }

                    // Spawn lightning based of player chance of it
                    if (chanceOfStrike != 0)
                    if (r.nextInt(100) < chanceOfStrike) {
                        map.getWorld().spawn(map.getWorld().getHighestBlockAt(l.getBlockX(), l.getBlockZ()).getLocation(), LightningStrike.class);
                    }

                    // Add creeper charged spawn if there are more players nearby
                    if (playersNearby > 0)
                        if (!creeperChargedSpawns.containsKey(l))
                            creeperChargedSpawns.put(l, playersNearby);

                }
            }

            if (creeperChargedSpawns.size() > 0) {
                for (Map.Entry<Location, Integer> entry : creeperChargedSpawns.entrySet()) {
                    if (r.nextInt(4) <= entry.getValue()) {
                        Location l = entry.getKey();
                        boolean spawned = false;
                        while (!spawned) {
                            Location temp = l.add(r.nextInt(7) - 3, 0, r.nextInt(7) - 3);
                            if (temp.getBlock().getRelative(BlockFace.UP).getType() == Material.AIR) {
                                Creeper c = map.getWorld().spawn(temp, Creeper.class);
                                c.setPowered(true);
                                spawned = true;
                            }
                        }
                    }
                }
                creeperChargedSpawns.clear();
            }


        }, startDelay, 20L);

    }
}

File: src/main/java/me/tcklpl/naturaldisaster/disasters/Thunderstorm.kt
package me.tcklpl.naturaldisaster.disasters;

import me.tcklpl.naturaldisaster.map.DisasterMap;
import me.tcklpl.naturaldisaster.util.BiomeUtils;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.entity.LightningStrike;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import java.util.Arrays;
import java.util.Random;
import java.util.concurrent.atomic.AtomicInteger;

public class Thunderstorm extends Disaster {

    public Thunderstorm() {
        super("Thunderstorm", true, Material.CREEPER_HEAD, BiomeUtils.PrecipitationRequirements.SHOULD_RAIN);
    }

    private void spawnLightningStrikeAt(Location loc) {
        LightningStrike ls = map.getWorld().spawn(loc, LightningStrike.class);
        Block b = ls.getLocation().getBlock();

        // break blocks around the lighting strike
        Arrays.stream(BlockFace.values()).map(b::getRelative).forEach(rel -> rel.breakNaturally(new ItemStack(Material.AIR)));
    }

    @Override
    public void startDisaster() {
        super.startDisaster();
        map.setPrecipitation(DisasterMap.MapPrecipitation.PRECIPITATE);

        Random r = random;

        AtomicInteger randomStrikesTimesRan = new AtomicInteger(0);
        AtomicInteger randomLightningStrikes = new AtomicInteger(1);

        // Spawn random lightning strikes on the map
        int randomLightningTask = Bukkit.getScheduler().scheduleSyncRepeatingTask(main, () -> {

            for (var i = 0; i < randomLightningStrikes.get(); i++) {
                Bukkit.getScheduler().scheduleSyncDelayedTask(main, () -> {

                    var loc = map.getRandomBlockInMap();
                    spawnLightningStrikeAt(loc);

                }, r.nextInt(3 * 20));
            }

            // Spawn 1 more lightning strike per cycle each 10s
            if (randomStrikesTimesRan.incrementAndGet() % 2 == 0) {
                randomLightningStrikes.incrementAndGet();
            }

        }, startDelay, 5 * 20);

        // Spawn lightning strikes somewhere around the players. The dispersion varies based on some conditions
        int lightningOnPlayerTask = Bukkit.getScheduler().scheduleSyncRepeatingTask(main, () -> {

            map.getPlayersInArena().forEach(p -> {
                int y = p.getLocation().getBlockY();
                int dispersion = 25;

                // Decrease the dispersion based on the player Y level
                int heightPenalityStep = Math.floorDiv(map.getMapSize().getY(), 10);
                int heightPenality = Math.floorDiv(y - map.getMinY(), heightPenalityStep);
                dispersion -= heightPenality;

                // Decrease the dispersion if there's no blocks above the players head
                if (y >= map.getWorld().getHighestBlockYAt(p.getLocation().getBlockX(), p.getLocation().getBlockZ())) {
                    dispersion -= 5;
                }

                // Decrease the dispersion in 1 for each player nearby
                var playersNearby = p.getNearbyEntities(3, 3, 3).stream().filter(e -> e instanceof Player).count();
                dispersion -= (int) playersNearby;

                // Spawn the lightning strike somewhere in a box centered around the player based on the dispersion
                int finalDispersion = Math.max(dispersion, 0);
                Bukkit.getScheduler().scheduleSyncDelayedTask(main, () ->
                        spawnLightningStrikeAt(map.getWorld().getHighestBlockAt(
                                p.getLocation().getBlockX() - finalDispersion + r.nextInt(2 * finalDispersion),
                                p.getLocation().getBlockZ() - finalDispersion + r.nextInt(2 * finalDispersion)
                        ).getLocation())
                , r.nextInt(2 * 20));
            });

        }, startDelay, 3 * 20);

        registerTasks(randomLightningTask, lightningOnPlayerTask);

    }
}
